### ISSUE TITLE
Engine crashes if pass an userdata as arg for standart lua error() function

### ISSUE THREAD

**Issue by alexeyfeskov** [defold/defold#9392]
**Describe the bug (REQUIRED)**
Defold engine crashes if provide an userdata as an argument for error() in lua 

**To Reproduce (REQUIRED)**
Create project from Empty Template, add sсript with `error(hash("error"))` in init()

**Expected behavior (REQUIRED)**
Expected: Console output `Error(userdata)`
Observed: Engine crashes with native platform error

**Defold version (REQUIRED):**
  - Version 1.9.2 and earlier

**Platforms (REQUIRED):**
tested on Win, HTML5 and Android - 100% reproducible

**Minimal repro case project (OPTIONAL):**
Describer above

**Logs (OPTIONAL):**
ERROR:CRASH: CALL STACK:

ERROR:CRASH:  0 0x7FF9439BB630 RaiseException <unknown>:0
ERROR:CRASH:  1 0x7FF7F0F904C0 wscanf_s <unknown>:0
ERROR:CRASH:  2 0x7FF7F0F91820 lj_err_throw <unknown>:0
ERROR:CRASH:  3 0x7FF7F0F91540 lj_err_run <unknown>:0
ERROR:CRASH:  4 0x7FF7F0F904C0 wscanf_s <unknown>:0
ERROR:CRASH:  5 0x7FF7F0F913B0 lj_err_optype <unknown>:0
ERROR:CRASH:  6 0x7FF7F0FA04A0 lj_meta_tget <unknown>:0
ERROR:CRASH:  7 0x7FF7F0F88840 lua_getfield <unknown>:0
ERROR:CRASH:  8 0x7FF7F0F5B230 dmScript::PCallInternal D:\a\defold\defold\engine\script\src\script.cpp:1407
ERROR:CRASH:  9 0x7FF7F0DAA820 dmGameObject::RunScript D:\a\defold\defold\engine\gameobject\src\gameobject\comp_script.cpp:144
ERROR:CRASH: 10 0x7FF7F0DA7D20 dmGameObject::CompScriptInit D:\a\defold\defold\engine\gameobject\src\gameobject\comp_script.cpp:185
ERROR:CRASH: 11 0x7FF7F0D8EC30 dmGameObject::InitComponents D:\a\defold\defold\engine\gameobject\src\gameobject\gameobject.cpp:1702
ERROR:CRASH: 12 0x7FF7F0D8E930 dmGameObject::InitCollection D:\a\defold\defold\engine\gameobject\src\gameobject\gameobject.cpp:1771
ERROR:CRASH: 13 0x7FF7F0D6DA70 dmEngine::Init D:\a\defold\defold\engine\engine\src\engine.cpp:1389
ERROR:CRASH: 14 0x7FF7F0D72740 dmEngineCreate D:\a\defold\defold\engine\engine\src\engine.cpp:2215
ERROR:CRASH: 15 0x7FF7F0D72CD0 dmEngine::RunLoop D:\a\defold\defold\engine\engine\src\engine_loop.cpp:68
ERROR:CRASH: 16 0x7FF7F0D6AD70 engine_main D:\a\defold\defold\engine\engine\src\engine_main.cpp:152
ERROR:CRASH: 17 0x7FF7F11B1614 __scrt_common_main_seh D:\a\_work\1\s\src\vctools\crt\vcstartup\src\startup\exe_common.inl:288
ERROR:CRASH: 18 0x7FF944517360 BaseThreadInitThunk <unknown>:0
ERROR:CRASH: 19 0x7FF945F1CC70 RtlUserThreadStart <unknown>:0

**Workaround (OPTIONAL):**
Do not pass userdata to lua error() function

**Comment by AGulev** [comment]
Doesn't it show 
>PANIC: unprotected error in call to Lua API (attempt to index a userdata value)
?

**Comment by AGulev** [comment]
that's what I get:
mac
![CleanShot 2024-09-14 at 10 04 27@2x](https://github.com/user-attachments/assets/d446d940-eada-40ca-a7b2-7a3516e862fd)

html
![CleanShot 2024-09-14 at 10 05 01@2x](https://github.com/user-attachments/assets/3ec0ae63-16d3-48ff-96c3-b21e71993120)

and this is expected, because it is Lua API and Lua requires string there.
I can't get crash you mentioned. The project I tested with is attached:
[ErrorHash.zip](https://github.com/user-attachments/files/17001528/ErrorHash.zip)

**Comment by alexeyfeskov** [comment]
> Please test my example
My bad, I was in a hurry to report a bug and didn't provide full
For HTML5 I've got pretty same behavior and output (PANIC)

But for debug Win build - got a CRASH (application immediately closes)
![image](https://github.com/user-attachments/assets/cd89559c-16a0-430b-9cd1-de04b031c2b4)
[ErrorHash-Win.log](https://github.com/user-attachments/files/17002416/ErrorHash-Win.log)
[_crash-Win.zip](https://github.com/user-attachments/files/17002417/_crash-Win.zip)

**Comment by alexeyfeskov** [comment]
On Debug Android ARM64 build - application immediately closes too
There is no any _crash file in `Android\data\com.example.todo\files` folder

**Comment by alexeyfeskov** [comment]
> this is expected, because it is Lua API and Lua requires string there
Just for me - this is NOT expected behavior. It's just a error() in my script - and I'm expect a error log and break the logic of my SINGLE script, but I didn't expect crashing whole application for native reasons
Especially for Android -  I can't get any clues that's going wrong, there is no _crash file, I can't find any logs using logcat..

**Comment by AGulev** [comment]
You don't see a crash file because it's not a crash. Lua Panic is emergency exit from application on Lua side. In logs it should be showns as:
>PANIC: unprotected error in call to Lua API (attempt to index a userdata value)

By "expected" I mean, it's internal Lua functionality and how Lua behaves in such cases. It's not something we handle on Defold side.

Usually, we don't patch Lua but I'm not sure what we should do in such case. @britzl @JCash any ideas? Similiar case with Lua Panic: https://github.com/defold/defold/issues/8540

**Comment by alexeyfeskov** [comment]
Do we need to report this as a bug to Lua (not sure about end-point address)?
> Although you can use a **value of any type** as an error message, usually error messages are strings describing what went wrong
[proof](https://www.lua.org/pil/8.5.html)

Lua is a part of Defold engine, and it's valuable and precious part. But there are many other values (I'm even can't count most of these) which do "out-of-box" solution for most of us. And I'll happy if we can find any decision: PANIC without logs it's not that we expect

**Comment by britzl** [comment]
> Usually, we don't patch Lua but I'm not sure what we should do in such case

Correct. We will not patch Lua. That is not going to happen. BUT is there nothing we can do to prevent the entire engine from crashing when there is an error such as this in a script?
